How to use the OpenMP 5.0 backend?

Hi,
I am trying to use the OpenMP4 backend with GCC 7.4 on Ubuntu 18.04, but it is not working.

I have modified `example/vectorAdd/src/vectorAdd.cpp` to use `AccCpuOmp4`:
```diff
diff --git a/example/vectorAdd/src/vectorAdd.cpp b/example/vectorAdd/src/vectorAdd.cpp
index b14f0198ba49..d9d47b5819fd 100644
--- a/example/vectorAdd/src/vectorAdd.cpp
+++ b/example/vectorAdd/src/vectorAdd.cpp
@@ -95,7 +95,7 @@ auto main()
     // - AccCpuOmp4
     // - AccCpuTbbBlocks
     // - AccCpuSerial
-    using Acc = alpaka::acc::AccCpuSerial<Dim, Idx>;
+    using Acc = alpaka::acc::AccCpuOmp4<Dim, Idx>;
     using DevAcc = alpaka::dev::Dev<Acc>;
     using PltfAcc = alpaka::pltf::Pltf<DevAcc>;
 
```

Then I tried building it with
```bash
mkdir example/vectorAdd/build
cd example/vectorAdd/build
cmake ..
make
```

`cmake` seems happy enough:
```log
-- The C compiler identification is GNU 7.4.0
-- The CXX compiler identification is GNU 7.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc - works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ - works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found TBB: /usr/include (found suitable version "2019.0", minimum required is "2.2")  
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Found CUDA: /usr/local/cuda (found suitable version "10.2", minimum required is "9.0") 
-- ALPAKA_ACC_CPU_B_SEQ_T_SEQ_ENABLED
-- ALPAKA_ACC_CPU_B_SEQ_T_THREADS_ENABLED
-- ALPAKA_ACC_CPU_B_TBB_T_SEQ_ENABLED
-- ALPAKA_ACC_CPU_B_OMP2_T_SEQ_ENABLED
-- ALPAKA_ACC_CPU_B_SEQ_T_OMP2_ENABLED
-- ALPAKA_ACC_CPU_BT_OMP4_ENABLED
-- ALPAKA_ACC_GPU_CUDA_ENABLED
-- Found alpaka: /home/fwyzard/src/alpaka/alpaka/include (found version "0.5.0") 
-- Configuring done
-- Generating done
-- Build files have been written to: /home/fwyzard/src/alpaka/alpaka/example/vectorAdd/build
```
but `make` fails with
```log
[ 50%] Building NVCC (Device) object CMakeFiles/vectorAdd.dir/src/vectorAdd_generated_vectorAdd.cpp.o
Scanning dependencies of target vectorAdd
[100%] Linking CXX executable vectorAdd
lto1: internal compiler error: bytecode stream: expected tag fixed_point_type instead of error_mark
Please submit a full bug report,
with preprocessed source if appropriate.
See <https://gcc.gnu.org/bugs/> for instructions.
mkoffload: fatal error: /usr/bin/x86_64-linux-gnu-accel-nvptx-none-gcc-7 returned 1 exit status
compilation terminated.
lto-wrapper: fatal error: /usr/lib/gcc/x86_64-linux-gnu/7//accel/nvptx-none/mkoffload returned 1 exit status
compilation terminated.
/usr/bin/ld: error: lto-wrapper failed
collect2: error: ld returned 1 exit status
CMakeFiles/vectorAdd.dir/build.make:99: recipe for target 'vectorAdd' failed
make[2]: *** [vectorAdd] Error 1
CMakeFiles/Makefile2:93: recipe for target 'CMakeFiles/vectorAdd.dir/all' failed
make[1]: *** [CMakeFiles/vectorAdd.dir/all] Error 2
Makefile:100: recipe for target 'all' failed
make: *** [all] Error 2
```

It looks like the compiler is trying to offload to the nvptx backend, and fails.

How do I make use of the OpenMP 4 backend to run the code on the CPU ?

--- a/README_OMP5.md
+++ b/README_OMP5.md
@@ -1,53 +1,14 @@
 # Building configuring for the OMP5 backend
 
-To make the build system enable the OpenMP5 backend, one has to tell CMake
-explicitly about the OpenMP version supported by the compiler. CMake does not
-determine it automatically for some compilers.
 ```
-cmake -DOpenMP_CXX_VERSION=5 \
+cmake \
   -Dalpaka_ACC_ANY_BT_OMP5_ENABLE=on \
   -DBUILD_TESTING=on \
   -Dalpaka_BUILD_EXAMPLES=on \
 ```
-All other backends are disable for faster compilation/testing and reduced
-environment requirements. Add flags to set the required compiler and linker flags, e.g:
-- clang/AOMP, target x86:
-  ```
-    -DCMAKE_CXX_FLAGS="-fopenmp -fopenmp=libomp -fopenmp-targets=x86_64-pc-linux-gnu" \
-    -DCMAKE_EXE_LINKER_FLAGS="-fopenmp"
-  ```
-- clang/AOMP, target ppc64le:
-  ```
-    -DCMAKE_CXX_FLAGS="-fopenmp -fopenmp=libomp -fopenmp-targets=ppc64le-pc-linux-gnu" \
-    -DCMAKE_EXE_LINKER_FLAGS="-fopenmp"
-  ```
-- clang, target nvptx:
-  ```
-    -DCMAKE_CXX_FLAGS="-fopenmp -fopenmp-targets=nvptx64-nvidia-cuda -O2" \
-    -DCMAKE_EXE_LINKER_FLAGS="-fopenmp"
-  ```
-- AOMP, target amdhsa:
-  ```
-    -DCMAKE_CXX_FLAGS="-fopenmp=libomp -fopenmp-targets=amdgcn-amd-amdhsa -Xopenmp-target=amdgcn-amd-amdhsa -march=gfx900 --save-temps" \
-    -DCMAKE_EXE_LINKER_FLAGS="-fopenmp"
-  ```
-- GCC, target nvptx:
-  ```
-    -DCMAKE_CXX_FLAGS="-foffload=nvptx-none -foffload=-lm -fno-lto"
-  ```
-- GCC, target host:
-  ```
-    -DCMAKE_CXX_FLAGS="-foffload=disable -fno-lto"
-  ```
-  - To run set the environment variable `OMP_TARGET_OFFLOAD=DISABLED`.
-- XL, offload:
-  ```
-    -DCMAKE_CXX_FLAGS="-qoffload -qsmp"
-  ```
-- XL, no offload:
-  ```
-    -DCMAKE_CXX_FLAGS=""
-  ```
+Also set `-DCMAKE_CXX_FLAGS` in accordance with compiler and target. Information
+about this is maintained
+[on readthedocs](https://alpaka.readthedocs.io/en/latest/advanced/cmake.html).
 
 ## Block-shared Memory
 
@@ -85,6 +46,7 @@ using the `alpaka_OFFLOAD_USE_BUILTIN_SHARED_MEM` flag:
 | gcc 11 | x86 | :white_check_mark: | N | N |
 | gcc 11 | nvptx | :white_check_mark:/:x: (4.) | N | N |
 | nvhpc 22.1 | x86 | :white_check_mark: | N (5.) | N (5.) |
+| nvhpc 22.3 | nvptx | :white_check_mark: | N (5.) | N (5.) |
 
 Keys:
 * :white_check_mark:: Test Passes.

--- a/docs/source/advanced/cmake.rst
+++ b/docs/source/advanced/cmake.rst
@@ -165,7 +165,7 @@ OpenMP 5
 alpaka_ACC_ANY_BT_OMP5_ENABLE
   .. code-block::
 
-     Enable the OpenMP 5.0 CPU block and block thread back-end.
+     Enable the OpenMP 5.0 any target block and block thread back-end.
 
 
 alpaka_OFFLOAD_MAX_BLOCK_SIZE
@@ -173,6 +173,41 @@ alpaka_OFFLOAD_MAX_BLOCK_SIZE
 
      Maximum number threads per block to be suggested by any target offloading backends
      ANY_BT_OMP5 and ANY_BT_OACC.
+
+
+CMAKE_CXX_FLAGS
+  Target architecture and some compiler specific flags have to be set manually:
+  
+  * Clang / AOMP / rocmClang
+    
+    * x86: `-fopenmp -fopenmp-targets=x86_64-pc-linux-gnu`
+      
+    * ppc64le: `-fopenmp -fopenmp-targets=ppc64le-pc-linux-gnu`
+      
+    * hsa: `-fopenmp -fopenmp-targets=amdgcn-amd-amdhsa -Xopenmp-target=amdgcn-amd-amdhsa -march=<arch>`
+      
+      A list of AMD GPU architectures can be found `here <https://llvm.org/docs/AMDGPUUsage.html#processors>`_.
+      
+    * nvptx: `-fopenmp -fopenmp-targets=nvptx64-nvidia-cuda`
+
+    CMake actually sets `-fopenmp` automatically, but OpenMP support is
+    detected after the deteced compiler is tested, which fails with
+    `-fopenmp-targets=` being present alone.
+      
+  * GCC
+    
+    * host: `-foffload=disable`
+      
+      At run time set the environment variable `OMP_TARGET_OFFLOAD=DISABLED`
+      
+    * nvptx: `-foffload=nvptx-none`
+      
+  * NVHPC
+    
+    * host: `-ta=host` or `-ta=-ta=multicore`
+      
+    * nvptx: `-ta=tesla`
+
 
 .. _cuda:
 

--- a/docs/source/advanced/cmake.rst
+++ b/docs/source/advanced/cmake.rst
@@ -191,7 +191,7 @@ CMAKE_CXX_FLAGS
     * nvptx: `-fopenmp -fopenmp-targets=nvptx64-nvidia-cuda`
 
     CMake actually sets `-fopenmp` automatically, but OpenMP support is
-    detected after the deteced compiler is tested, which fails with
+    detected after the detected compiler is tested, which fails with
     `-fopenmp-targets=` being present alone.
       
   * GCC
@@ -204,7 +204,7 @@ CMAKE_CXX_FLAGS
       
   * NVHPC
     
-    * host: `-ta=host` or `-ta=-ta=multicore`
+    * host: `-ta=host` or `-ta=multicore`
       
     * nvptx: `-ta=tesla`
 

